docs/book/v2/routing-middleware.md

```diff
@@ -0,0 +1,141 @@
+                                RequestHandlerInterface $handler
+                            ): ResponseInterface {
+                                // Attempt to use $handler here will always result in empty pipeline exception 
+                            }
+                        },
+                    ],
+                ],
+            ],
+            'instance-handler' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-handler',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new class implements RequestHandlerInterface {
+                            public function handle(
+                                ServerRequestInterface $request
+                            ): ResponseInterface {
+                                //  
+                            }
+                        },
+                    ],
+                ],
+            ],
+            'instance-pipespec-preferred' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-pipespec-preferred',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new PipeSpec(
+                            'container_id_for_middleware',
+                            'container_id_for_another_middleware',
+                            'container_id_for_handler',
+                        ),
+                    ],
+                ],
+            ],
+            'instance-pipespec-dev-env' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-pipespec-dev-env',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new PipeSpec(
+                            'container_id_for_middleware',
+                            new class implements MiddlewareInterface {
+                                public function process(
+                                    ServerRequestInterface $request,
+                                    RequestHandlerInterface $handler
+                                ): ResponseInterface {
+                                    return $handler->handle($request);
+                                }
+                            },
+                            static function (
+                                ServerRequestInterface $request,
+                                RequestHandlerInterface $handler
+                            ): ResponseInterface {
+                                return $handler->handle($request);
+                            },
+                            'container_id_for_handler',
+                        ),
+                    ],
+                ],
+            ],
+        ],
+    ],
+];
```

Comment:
The whole code example should be split and each example should have its own heading.